Make GenerateEvents imply inlining.

HansOlsson opened the pull request:
Closes #2986

henrikt-ma commented on chapters/functions.tex:
```diff
@@ -1883,7 +1883,7 @@ \section{Function Inlining and Event Generation}\label{function-inlining-and-eve
 
 If \lstinline!GenerateEvents = true!, the model developer proposes that crossing functions in the function should generate events (one possibility of doing this is to inline the function and generate events for the inlined function).
-
+As defaults this implies \lstinline!Inline = true!, but it can be overriden by specifying one of the inline-annotations.
```
I don't think this is ready for a proper review, but I can already note that this doesn't rule out combining `GenerateEvents = true` with `Inline = false`, which I thought was generally perceived as an undesired combination.

eshmoylova replied:
I remember asking this question in the meeting but now I don't remember the answer: What happens if GenerateEvents = true and the function cannot be inlined?

henrikt-ma replied:
This is precisely the reason why `GenerateEvents = true` should imply inlining, meaning that it is a tool failure if it can't be inlined.  Put differently, the point is that tools are not expected to be able to handle inlining by any other means than by inlining.

henrikt-ma replied:
This conversation is obsolete in view of later conversations and current state of the changes in this PR.  Resolving.

henrikt-ma commented on chapters/functions.tex:
```diff
@@ -1883,7 +1883,7 @@ \section{Function Inlining and Event Generation}\label{function-inlining-and-eve
 Has only an effect within a function declaration
```
Could we fix the missing punctuation here while we're at it?
```suggestion
Has only an effect within a function declaration.
```

HansOlsson commented:
I removed the problematic sentence and added two parts:

- Overriding with other inline annotations should have them equal to true (so you inline in other ways; don't disable it).
- If events cannot be generated diagnostics should be recommended.

Note that if a tool figures out a way of generating events without inlining that would still be allowed; as the requirement is to generate events - not to inline.

henrikt-ma commented:
> Note that if a tool figures out a way of generating events without inlining that would still be allowed; as the requirement is to generate events - not to inline.

Sure, but we also need to take into account what it reasonable to expect of tools.  By making it clear that any `GenerateEvents  = true` _can_ be handled using inlining, we show tool makers that it is not necessary (as far as the specification is concerned) to also consider more advanced implementations that don't rely on inlining.

henrikt-ma commented on chapters/functions.tex:
```diff
@@ -1880,10 +1880,11 @@ \section{Function Inlining and Event Generation}\label{function-inlining-and-eve
+Has only an effect within a function declaration.
 
+If \lstinline!GenerateEvents = true!, the model developer proposes that crossing functions in the function should generate events.
+As defaults this implies \lstinline!Inline = true!, but it can be overriden by specifying one of the inline-annotations (with true value).
```
I don't see clear support for allowing `LateInline` in the discussion yet, so either we need to also work on the definition of `LateInline`, or not include it for now:
```suggestion
This implies \lstinline!Inline = true! unless overridden by \lstinline!InlineAfterIndexReduction = true! (in particular, \lstinline!GenerateEvents = true! cannot be combined with \lstinline!Inline = false!).
```

Also note spelling of _overridden_.

eshmoylova replied:
Here is something else that we should consider. All of those annotations (various Inlines and GenerateEvents) are formulated as "the model developer proposes..." To me it implies, that a tool is not required to follow any of those suggestions. And for inlining it makes sense: the result of the evaluation should be the same. For events, it actually changes the result. So does it make sense to keep it as "model developer proposes"?

casella replied:
I fully agree with @eshmoylova's comment, and I would change "proposes" to "requires". That's the whole point of the `GenerateEvents` annotation: the modeller intends that events are indeed generated.

henrikt-ma replied:
Let us please continue this discussion either at line 1885 or 1887.  The suggestion for line 1886 has nothing to do with _proposes_ vs _requires_.

HansOlsson replied:
I still don't see the problem with implementing LateInline, but I think the other change of GenerateEvents make this redundant, as the basic requirement is to generate events. So if a tool cannot generate events if inlining that late the alternative is to inline earlier.

henrikt-ma replied:
I agree that the situation is better now that `GenerateEvents` is imperative (still needs decision) while the inlining annotations are merely propositions.  Still, the combination with `LateInline` will appear problematic in tools where this corresponds to an inlining stage when it is too late to process `GenerateEvents = true`.  A non-normative remark could clarify two things:
- That tools should override any given inlining annotation that it can't respect in combination with `GenerateEvents = true`, instead performing the inlining at the nearest stage where `GenerateEvents = true` can still be respected.
- That a diagnostic is recommended in case inlining isn't performed at the stage proposed by the inlining annotation(s).

henrikt-ma replied:
I just realize that this happens to be the place where the non-normative text should go, so here's a (un-tested) suggestion:
```suggestion
This implies \lstinline!Inline = true! unless overridden by specifying one of the inlining annotations with value \lstinline!true! (in particular, \lstinline!GenerateEvents = true! cannot be combined with \lstinline!Inline = false!).

\begin{nonnormative}
In case a given inlining annotation proposes to inline at a stage when it is too late for a tool to process \lstinline!GenerateEvents = true!, the tool should instead perform inlining of the function at the nearest stage where \lstinline!GenerateEvents = true! can still be respected.
In this case, a diagnostic is recommended.
\end{nonnormative}
```

eshmoylova replied:
> * That tools should override any given inlining annotation that it can't respect in combination with `GenerateEvents = true`, instead performing the inlining at the nearest stage where `GenerateEvents = true` can still be respected.

I like this suggestion because it accounts for the different tools having different abilities without making one or the other illegal. Is it worth mentioning that GenerateEvents =true may result in different results in different tools or is it obvious?

Also the suggested change above has a typo: "instead instead".

henrikt-ma replied:
> Also the suggested change above has a typo: "instead instead".

Thanks, fixed.

henrikt-ma replied:
> Is it worth mentioning that GenerateEvents =true may result in different results in different tools or is it obvious?

I don't find it obvious, especially considering the effort we've made to avoid ambiguity due to uncertain application of inverse functions.  However, I suppose we could construct an example where `GenerateEvents = true` could make index reduction fail by inlining before index reduction and thereby making a `derivative` unavailable?

HansOlsson replied:
It looks good. Except I will remove 'too late' as there could be tools having problem with doing it too early.

henrikt-ma commented on chapters/functions.tex:
```diff
@@ -1880,10 +1880,11 @@ \section{Function Inlining and Event Generation}\label{function-inlining-and-eve
 
+If \lstinline!GenerateEvents = true!, the model developer proposes that crossing functions in the function should generate events.
+As defaults this implies \lstinline!Inline = true!, but it can be overriden by specifying one of the inline-annotations (with true value).
+Tools should generate diagnostics if it is not possible to generate events for crossing functions in the function.
```
Trying to formulate in a more familiar (but stronger than usual) way:
```suggestion
As event-generation can be crucial for correct simulation, a diagnostic is strongly recommended in case events will not be generated for a function with `GenerateEvents = true`.
```

Alternatively, wouldn't it be better if we took the opportunity to remove the tool's freedom to choose whether to generate events or not?

henrikt-ma reviewed (changes requested):
Seems like we might have to do some serious work on `LateInline` before we can finish this.

HansOlsson commented:
> > Note that if a tool figures out a way of generating events without inlining that would still be allowed; as the requirement is to generate events - not to inline.
> 
> Sure, but we also need to take into account what it reasonable to expect of tools. By making it clear that any `GenerateEvents = true` _can_ be handled using inlining, we show tool makers that it is not necessary (as far as the specification is concerned) to also consider more advanced implementations that don't rely on inlining.

That was exactly the intent. GenerateEvents can only be combined with some variant of inlining enabled; and tools should generate a warning if they cannot generate crossing functions to handle events.

henrikt-ma commented on chapters/functions.tex:
```diff
@@ -1880,10 +1880,11 @@ \section{Function Inlining and Event Generation}\label{function-inlining-and-eve
 
+If \lstinline!GenerateEvents = true!, the model developer proposes that crossing functions in the function should generate events.
+As defaults this implies \lstinline!Inline = true!, but it can be overriden by specifying one of the inline-annotations (with true value).
+As event-generation can be crucial for correct simulation, a diagnostic is strongly recommended in case events will not be generated for a function with `GenerateEvents = true`.
```
As noted by @eshmoylova and @casella (and I totally agree), it would be great if we could simply drop this line and instead make `GenerateEvents` imperative.

HansOlsson replied:
Ok

HansOlsson replied:
I think that is a good to make it clear that GenerateEvents shall truly generate events.

henrikt-ma replied:
As the description of `GenerateEvents = true` now says _shall_, this sentence should be removed, and then we can resolve this conversation as long as we remember to bring up the change from _proposes_ to _shall_ for a language group decision.

HansOlsson replied:
Ok

henrikt-ma commented on chapters/functions.tex:
```diff
@@ -1880,10 +1880,11 @@ \section{Function Inlining and Event Generation}\label{function-inlining-and-eve
-If \lstinline!GenerateEvents = true!, the model developer proposes that crossing functions in the function should generate events (one possibility of doing this is to inline the function and generate events for the inlined function).
+Has only an effect within a function declaration.
 
+If \lstinline!GenerateEvents = true!, the model developer proposes that crossing functions in the function should generate events.
```
What @eshmoylova, @casella and I propose would be to remove the tool's freedom to respect `GenerateEvents` or not, something like this:
```suggestion
By default, \lstinline!GenerateEvents = false! and expressions in the function body that would normally be event-generating shall not generate events, similar to inlining the function body while wrapping all expressions in \lstinline!noEvent!, see \cref{modelica:noEvent}.
By specifying \lstinline!GenerateEvents = true!, event-generating expressions in the function body shall generate events as normal, similar to inlining the function body without wrapping all expressions in \lstinline!noEvent!.
```

henrikt-ma reviewed (changes requested):
Appart from the suggestion made above, the main obstacle the I see remaining here is that we need a language group decision on the change from _proposes_ to _shall_.  Marking as _Requests changes_ just as a reminder of that.

HansOlsson commented on chapters/functions.tex:
```diff
@@ -1883,7 +1883,7 @@ \section{Function Inlining and Event Generation}\label{function-inlining-and-eve
 Has only an effect within a function declaration
 
 If \lstinline!GenerateEvents = true!, the model developer proposes that crossing functions in the function should generate events (one possibility of doing this is to inline the function and generate events for the inlined function).
```
It has been updated.

HansOlsson commented:
I believe all the issues are now resolved. Due to the new strictness I also added that `GenerateEvents=true` doesn't matter if used where events will not be generated. (Since enclosing or not enclosing the result in `noEvent `doesn't matter in those cases.)

eshmoylova commented on chapters/functions.tex:
```diff
@@ -1880,10 +1880,17 @@ \section{Function Inlining and Event Generation}\label{function-inlining-and-eve
+
+By default, \lstinline!GenerateEvents = false! and expressions in the function body that would normally be event-generating shall not generate events, similar to inlining the function body while wrapping all expressions in \lstinline!noEvent!, see \cref{modelica:noEvent}.
+By specifying \lstinline!GenerateEvents = true!, event-generating expressions in the function body shall generate events as normal, similar to inlining the function body without wrapping all expressions in \lstinline!noEvent!.
+This implies \lstinline!Inline = true! unless overridden by specifying one of the inlining annotations with value \lstinline!true! (in particular, \lstinline!GenerateEvents = true! cannot be combined with \lstinline!Inline = false!).
```
It is not clear what "This" is: if it refers to both sentences above or just the last one. How about this?
```suggestion
In particular, \lstinline!GenerateEvents = true! implies \lstinline!Inline = true! unless overridden by specifying one of the inlining annotations with value \lstinline!true! (in particular, \lstinline!GenerateEvents = true! cannot be combined with \lstinline!Inline = false!).
```

henrikt-ma replied:
I see the problem with _this_, but also don't quite like _in particular_ as the way of connecting the two sentences.  To me, _in particular_ makes it sound as if the second sentence is a consequence of the first one, rather than a separate statement.  I'd therefore suggest:
```suggestion
To simplify implementation, \lstinline!GenerateEvents = true! implies \lstinline!Inline = true! unless overridden by specifying one of the inlining annotations with value \lstinline!true! (in particular, \lstinline!GenerateEvents = true! cannot be combined with \lstinline!Inline = false!).```

eshmoylova replied:
@henrikt-ma, what are ``` at the end of your suggestions for? Looks like GitHub markup.

I wasn't sure how to connect the two sentences because I wasn't sure what *this* in the original sentence was referring to. It looked to me that the explanation presented in the sentence before this one was the one that implied that GenerateEvents = true should imply that.... So, *in particular* or *thus* or *therefore* would have fit with that idea. I am also not sure about *To simplify implementation*. To me it suggests that if you don't have an implementation that's what you should do. I thought the agreement was that GenerateEvents = true, a tool will try to inline to be able to generate the events. If inlining is not possible then the tool either tries a more sophisticated algorithm if it has one or issues a diagnostic. With *To simplify implementation* it looks to me more like if you get GenerateEvents = true and you have a sophisticated implementation without inlining, you do not need to try to inline it. I don't know if this distinction matters. Probably the end result will be the same if you generate events by inlining or by some other means. But I want to bring it up to confirm that it's acceptable and is consistent with the group agreement.

HansOlsson replied:
Hm... I don't see why we need to mention implementations here, or need to use 'This' to reference something.
So what about just saying the minimum: "Having \lstinline!GenerateEvents = true! implies \lstinline!Inline = true! ..."

eshmoylova replied:
Yes

HansOlsson replied:
Done now.

henrikt-ma replied:
> @henrikt-ma, what are ``` at the end of your suggestions for? Looks like GitHub markup.

(This was due to a missing newline at the end of the suggestion.  Fixed now.)

eshmoylova reviewed (commented):
One minor clarification on wording. Also there were comments that it should be reviewed/voted on by the Specification group given that the wording is much stronger. I am not sure what is the right way to submit my review in this case. Is it best to wait with approval until we have the decision? Or is it fine to approve now under the condition of the group approval? In the latter case, I'm afraid it would be easy to forget that it still needs to be approved by the group.

henrikt-ma reviewed (approved):
Approving, based on the assumption that we will actually have a language group discussion about this tomorrow.
